### ISSUE TITLE
Low precedence defined or - dor

### ISSUE THREAD

**Issue by caribpa** [Perl/perl5#17810]
I'm reading the Perl Cookbook at the moment and I was a little surprised to learn that the low precedence defined or (proposed as `err` or `dor`) never made it past [Perl 5.9](https://metacpan.org/pod/release/RGARCIA/perl-5.9.4/lib/feature.pm#the-'err'-feature) without any apparent reason.

I found this [Perl Monks](https://www.perlmonks.org/?node_id=660524) discussion that points to a long mailing list [thread](http://www.nntp.perl.org/group/perl.perl5.porters/2007/09/msg128170.html) which more or less concludes that the `err` keyword was ambiguous and confusing to newcomers, but what about `dor`? `dor` seems to be the right keyword.

Perl Monks also added that this operator was included in Raku as [orelse](https://docs.raku.org/routine/orelse) (also doing `$_` aliasing).

In my opinion the name `orelse` is more confusing than `dor`, but ignoring this personal preference, is there any reason why this operator was never (re-)added to Perl?

Currently under `use warnings` a warning is issued here:

```perl
use warnings;

my $n = <DATA> or die;  # Value of <HANDLE> construct can be "0"; test with defined()
print $n;

__DATA__
test
```

but not with `getc`:

```perl
use warnings;

my $n = getc DATA or die;  # No warning, it just dies
say $n;

__DATA__
0123
```

So that you are forced to use `defined()` when `dor` would be cleaner without the edge cases you can find with `//` due to its precedence:

```perl
use warnings;

defined(my $n = <DATA>) or die;  # No warning but less readable
print $n;

my $p = <DATA> // die;           # No warning and for this case you can forgo the parentheses 
print $p;

($p = <DATA>) // $n = 'test';    # But you'll have to remember the parentheses when the precedence requires it,
                                 # and for both parts or: Can't modify defined or (//) in scalar assignment
print $p;

my $d = <DATA> dor die;          # More readable
print $d;

$d = <DATA> dor $n = 'test';     # And does what you want
print $d;

__DATA__
test1
test2
test3
test4
test5
```

**Comment by xsawyerx** [comment]
I definitely don't see `dor` as a suitable op name.

I also don't see the value in hoffmanizing as much of the user code as possible. It's great for one-liners but this shouldn't be a priority for us or an interest of ours.

```perl
defined( my $d = <DATA> )
    or die;
```

...is perfectly fine to me. It's more readable and does the job. I don't have to explain this. `dor` I would have to explain and would rely on the person understanding the operator's precedence. knowledge which should be relied on sparingly.

**Comment by Tux** [comment]
I was/am i big proponent of low-precedence defined-or, be it `dor` or `err`.
I miss it on a daily basis.
I find it easier to explain (and read)
```perl
my $foo = function () err exit 0;
```
than to explain the scope of `$foo` in
```perl
defined (my $foo = function ()) or exit 0;
```
Perception and experience differ. I however won't pursue this.

**Comment by hvds** [comment]
I'm with @Tux on this, though I might not use it quite as often as daily.

**Comment by xsawyerx** [comment]
> I was/am i big proponent of low-precedence defined-or, be it `dor` or `err`.
> I miss it on a daily basis.
> I find it easier to explain (and read)
> 
> ```perl
> my $foo = function () err exit 0;
> ```
> 
> than to explain the scope of `$foo` in
> 
> ```perl
> defined (my $foo = function ()) or exit 0;
> ```
> 
> Perception and experience differ. I however won't pursue this.

Objectively, the parenthesis and `or` in `defined (...) or die` makes it a clearly parsed sentence without knowing what `die` does (though it's a big hint). Operators with exceptionally low precedence add an additional piece of knowledge required because otherwise, you cannot parse it in your head.

I think `err` as a function name is not bad. But as an `op` that also contains a condition requiring precedence knowledge is problematic because it's hidden and not available in the name. The word `err` also does not reflect anything that mentions a condition on it. (`assert` is probably better because the word "assert" notes a condition that must be met.)

**Comment by Tux** [comment]
I personally did not care about the operator-name. Even if it were called `defined_or`, I'd use it.
It is all about the ortogonality of the language. TIMTOWTDI. `err` and `dor` where work-names in those days. (I had a patch ready for all 5.8.x's that implemented it BTW)
I gave up on this late in 5.8.x and I won't put any effort in trying to get it in. The people that were in favor lost the battle when it was faught.

**Comment by caribpa** [comment]
@xsawyerx I understand that this may not be a really important feature but apart from being convenient, I also see it as a consistency because there are low precedence operators for every other logical operator.

I see that some people don't like these low precedence operators claiming that they make the code less readable, but there are also people like @Tux, @hvds, and me that happens to experience the exact opposite.

In my case I find the code cleaner with `dor` because there are less keywords (no need to write `defined`), less parenthesis pollution, and the indentation is preserved when defining multiple variables:

```perl
my $var1 = ...;
my $var2 = ...;
my $row1 = <> dor die;           # Same indentation and cleaner definition
(my $row2 = <>) // die;          # than this
defined (my $row3 = <>) or die;  # and definitely this
...
```

Fortunately nobody is forced to use such operators (thanks to TMTOWTDI) and you can even place some standards at your workplace/project (and create your own pragma/module that enforces them if you are so serious about it) so you don't have to waste time explaining nor debugging them.

And it shouldn't be a pain to add this feature because it was already implemented and functional in [Perl 5.9](https://metacpan.org/pod/release/RGARCIA/perl-5.9.4/lib/feature.pm#the-'err'-feature).

Regarding the naming, I find `dor` (d[efined] or) better than `err` and [Raku's](https://docs.raku.org/routine/orelse) `orelse` because of its mnemotechnics, because it *may* be used in a non-error/assert condition, and because we don't want to create confusion with `else` control flow statements. Anyways the name is not that important for me.

**Comment by Grinnz** [comment]
My vote is for defined_or. `err` does not provide any hint as to what it means, and `dor` seems nice but has no meaning, unlike every other low precedence version. Unfortunately it needs to be written out to have a clear and intuitive meaning like the others. I also believe it's still worth adding.

**Comment by Grinnz** [comment]
`def_or` would be enough to explain it as well, perhaps there are other compromises.

**Comment by caribpa** [comment]
I guess that back when the operator was proposed the suggested names had only 3 characters like some of the other operators.

Maybe `dor` was chosen following the naming logic of the `xor` operator, and if we are honest with ourselves `xor` has the same resemblance to Exclusive OR than `dor` to Defined OR (same can be said about the non self-explaining `cmp`, `<=>`, and the other comparison operators to some extent), so let's not pretend this is something different.

Also using a long name may well mislead some people into thinking that it is a function or a built-in instead of an operator.

Anyways I really like the `def_or`, so +1 @Grinnz

**Comment by Grinnz** [comment]
> Maybe `dor` was chosen following the naming logic of the `xor` operator, and if we are honest with ourselves `xor` has the same resemblance to Exclusive OR than `dor` to Defined OR

This is true but it's very different because `xor` is a well known abbreviation for this concept, while `dor` is purely an invention of these circumstances.

**Comment by caribpa** [comment]
> This is true but it's very different because `xor` is a well known abbreviation for this concept, while `dor` is purely an invention of these circumstances.

Agreed, but most probably the `xor` abbreviation was also an invention of the circumstances back then (I'm happy to be corrected about the origins of this abbreviation as a quick search did not yield anything interesting)

**Comment by xsawyerx** [comment]
Based on the pumpking for 5.10, `dor` and `err` were removed because "their usefulness was deemed low." I was wondering why we had them in 5.9.x but they weren't in 5.10.0.

I thought about this further and my main concern is that this seems like an arbitrary attempt to reduce the amount of code, but based on personal preference. It's fairly arbitrary.

For example, `//` seems incredibly useful and a lot of people use it. I personally rarely do. I use `defined and` much more than `defined or`.

```perl
defined $args{'get_bar'}
    and $opts{'bar'} = $foo->bar();
```

You could theoretically use the less readable `()!!x` but it has an unintentional effect:

```perl
my %opts = (
    ( 'bar' => $foo->bar() )x!! $args{'get_bar'}, # calls $foo->bar() anyway
);
```

So this leads me to several questions:

* Is symmetry a good enough reason to introduce this op?
* Why not introduce the corollary `defined and` op?
* Why not introduce the helpful `length and`?
* For that matter, why not `length or`?

It just seems arbitrary. Additionally, out of the list above, this particular op was available and removed, which requires going back to something we tried and decided was not worth it. If anything, it would merit additional research instead of being more reliable to add.

**Comment by rsFalse** [comment]
Some thoughts on it (may repeat).
* `defined_or` looks readable (self-explaining), but: 1. it has long name when it is supposed to be used in one-line statements; one needs more often to break lines, so he may prefer to choose `if(){}` instead; 2. it contains underscore, breaking a "tradition" that Perl operators (and most keywords) do not contain underscores; this might negatively influence parsing intuition;
* `dor` seems handy and suitable by its length (similar length to similar operators: `or`, `and`), but: 1. it is not-self-explained; 2. is a new name; 3. appears similar to `xor` but differs by returned value and ability to short-circuit; 4. looks similar to logical operators but isn't logical itself because short-circuits with boolean False.

**Comment by Tux** [comment]
I am in favor of getting `dor` back, but that will be no surprise
I am also in favor od getting `dand`, `land`, and `lor` or whatever they will be called (even if that needs a feature)

**Comment by caribpa** [comment]
Not to open old wounds, but I am genuine curious to know what was the reason for the removal of `dor` (and subsequent not addition of `dand`, `land`, `lor`, etc), the name, low usability?

I can understand that some people don't want to add more (possible) confusion to the code, but Perl itself has already way too many "magic" basic keywords (`$@`, `$_`, etc) that makes the code unreadable for beginners (who also complain about the use of semicolons, btw), so I personally don't see how `dor`/`defined_or` and the other ones suggested would be so harmful regarding readability.

Of course, these days my main usage of Perl is in one-liners (that end up too big), so certainly I appreciate compactness over readability.
